### PR TITLE
Add subscription model 

### PR DESCRIPTION
**Pull request by 0xkarmacoma**

### CODE REVIEW

File: src/OPTokenClaim.sol [hunk 1]
```diff
@@ -54,62 +70,139 @@ contract OPTokenClaim is Ownable {
...
-        require(
-            OP.transferFrom(treasury, account, claimableOP),
-            "Transfer failed"
-        );
+        require(OP.transferFrom(treasury, account, OPReward), "Transfer failed");
+
+        emit OPClaimed(account, lastEpochNum, OPReward);
 
-        emit OPClaimed(account, epoch, claimableOP);
+        if (epochNum < _config.maxEpoch) {
```
Comment (mariobyn) [inline src/OPTokenClaim.sol]: this should be above OP.transferFrom, keeping a CEI pattern.

File: src/OPTokenClaim.sol [hunk 2]
```diff
@@ -54,62 +70,139 @@ contract OPTokenClaim is Ownable {
...
-        if (expBalance > 99 ether) {
-            expBalance = 99 ether;
-        }
+        unchecked {
+            uint256 totalReward = totalEXP * 5 - 4 ether * subscribedAccounts;
 
-        return (expBalance * 5 - 4 ether);
-    }
+            // if total reward of given epoch is greater than 10k OP (MAX_REWARD), reduce reward
+            uint256 factor = 1 * 10 ** 18;
```
Comment (mariobyn) [inline src/OPTokenClaim.sol]: maybe use 1 ether instead of 1 * 10**18?

File: src/OPTokenClaim.sol (hunk 1 above)
Comment (0xkarmacoma) [reply src/OPTokenClaim.sol]: in [dcede7a](https://github.com/ethernautdao/ethernaut-op-claim/pull/3/commits/dcede7a0b9eb78f7763e975b22aa517e15855bff)

File: src/OPTokenClaim.sol [hunk 3]
```diff
@@ -54,62 +70,146 @@ contract OPTokenClaim is Ownable {
...
+        EpochInfo storage epoch = epochs[epochNum];
+        uint256 totalEXP = epoch.totalEXP;
+        uint256 subscribedAccounts = epoch.numAccounts;
 
-        // cap at 99 EXP
-        if (expBalance > 99 ether) {
-            expBalance = 99 ether;
-        }
+        unchecked {
+            uint256 totalReward = totalEXP * 5 - 4 ether * subscribedAccounts;
```
Comment (0xkarmacoma) [inline src/OPTokenClaim.sol]: looking closer at the math, I think there's a mistake here 

`(totalEXP * 5) - (4 ether * subscribedAccounts)`

vs 

`(totalEXP * 5 - 4 ether) * subscribedAccounts`

cc @beskay, let's wait to hear from @Dhannte to see if we can simplify the formula. I'd be in favor of getting rid of the `- 4` thing entirely

Comment (beskay) [reply src/OPTokenClaim.sol]: The first equation is correct, for example, we have 21 subscribed accounts with 99 EXP each:
totalEXP = 99 * 21 = 2079 EXP
total OP we have to distribute: (99 * 5 - 4) * 21 = 10311 OP

Above equation:
(2079 * 5) - 4 * 21 = 10311 OP

Im fine with simplifying the formula though.

File: src/OPTokenClaim.sol (hunk 2 above)
Comment (0xkarmacoma) [reply src/OPTokenClaim.sol]: simplified in 2d3841682d14ebc5ca8a33429924c3e9c2ab6d86

**Review by beskay — APPROVED** [review]
looks good!

File: src/OPTokenClaim.sol [hunk 4]
```diff
@@ -9,42 +9,48 @@ contract OPTokenClaim is Ownable {
...
         uint128 maxEpoch;
     }
 
     Config public config;
 
-    // epochNumber -> address -> claimed
-    mapping(uint256 => mapping(address => bool)) public epochToAddressClaimed;
+    mapping(uint256 => uint256) public totalEXPAtEpoch;
+
+    mapping(uint256 => mapping(address => uint256)) public epochToSubscribedEXP;
```
Comment (mariobyn) [inline src/OPTokenClaim.sol:30]: this should have a public getter method maybe?

File: src/OPTokenClaim.sol [hunk 5]
```diff
@@ -9,42 +9,48 @@ contract OPTokenClaim is Ownable {
...
 
         config = Config({
             start: 1669852800, // Thu Dec 01 2022 00:00:00 UTC
-            maxEpoch: 6
+            maxEpoch: 6 // first claim duration runs for 6 months
         });
 
-        // TODO: transferOwnership to EthernautDAO multisig?
+        // transfer ownership to multisig
+        _transferOwnership(treasury);
```
Comment (mariobyn) [inline src/OPTokenClaim.sol]: a small gas op, use `_treasury`

File: src/OPTokenClaim.sol [hunk 6]
```diff
@@ -54,62 +60,107 @@ contract OPTokenClaim is Ownable {
...
-            "already claimed for this epoch"
-        );
+        // cap at MAX_EXP = 99 EXP
+        if (expBalance > MAX_EXP) {
+            expBalance = MAX_EXP;
+        }
 
-        uint256 claimableOP = _calcReward(account);
+        uint256 subscribedEXP = epochToSubscribedEXP[epochNum][account];
+        if (subscribedEXP == expBalance) {
```
Comment (mariobyn) [inline src/OPTokenClaim.sol:110]: simply return is weird, i would revert with an error or emit an event at least

File: src/OPTokenClaim.sol [hunk 7]
```diff
@@ -54,62 +60,107 @@ contract OPTokenClaim is Ownable {
...
+            // calculate the total reward of given epoch
+            uint256 totalReward = totalEXPAtEpoch[epochNum] * 5;
+
+            // calculate individual reward
+            uint256 subscribedEXP = epochToSubscribedEXP[epochNum][account];
+            if (totalReward > MAX_REWARD) {
+                reward = 5 * subscribedEXP * MAX_REWARD / totalReward;
+            } else {
+                reward = 5 * subscribedEXP;
+            }
```
Comment (mariobyn) [inline src/OPTokenClaim.sol:87]: if totalReward == 95
user subscribed with 99 EXP
the reward to the user will be 5*99, is that ok?
Can we be in this situation?

File: src/OPTokenClaim.sol (hunk 4 above)
Comment (0xkarmacoma) [reply src/OPTokenClaim.sol:30]: there is the auto-generated public getter method, e.g.:

        // should have subscribed for epoch 5 by now
        assertEq(claimContract.epochToSubscribedEXP(5, alice), 10 ether);

File: src/OPTokenClaim.sol (hunk 6 above)
Comment (0xkarmacoma) [reply src/OPTokenClaim.sol:110]: in 4840dba

File: src/OPTokenClaim.sol (hunk 5 above)
Comment (0xkarmacoma) [reply src/OPTokenClaim.sol]: in 7d4bb60

File: src/OPTokenClaim.sol (hunk 7 above)
Comment (0xkarmacoma) [reply src/OPTokenClaim.sol:87]: do you mean `totalReward == 495`?

if a user subscribed with 99 EXP total reward is 495, that means no other user subscribed. So in that case the user would just get his share normally, nothing weird

Comment (mariobyn) [reply src/OPTokenClaim.sol:87]: total reward can never be less than 495? i did not do the math yet

Comment (0xkarmacoma) [reply src/OPTokenClaim.sol:87]: `totalReward` is `5 * totalEXP`

`totalEXP` is the sum of all `subscribedEXP`, so this should be true for all accounts:

    totalEXP >= subscribedEXP(account)

**Pull request merged**